### DISCUSSION TITLE
Help needed with background jobs and context injection

### DISCUSSION THREAD

**Discussion by franck102** [quarkusio/quarkus#33975] [Q&A]
Hi all, I need to run some background tasks on startup and those tasks need to access request scoped beans and contextual data. My current approach kind of works, but I am pretty sure that there must be a better / simpler way to achieve this.

My background tasks are started like this:
```
     ...
    @Inject
    RoutingContext _routingContext;

    @Inject
    Vertx _vertx;

    void onStart(@Observes StartupEvent ev)
    {
        Context context = _vertx.getOrCreateContext();
        if (context.isEventLoopContext()) {
            context = ((ContextInternal) context).duplicate();
        }
        for (String tenantId : _coreStore.listTenants()) {
            context.executeBlocking(call -> {
                try {
                    startGateway(tenantId);
                    call.complete();
                }
                catch (Throwable t) {
                    _log.error("Could not start payment gateway for tenant " + tenantId, t);
                    call.fail(t);
                }
            });
        }
    }

    @ActivateRequestContext
    public void startGateway(String tenantId)
    {
        // _routingContext.put(CoreUtils.KEY_TENANT_ID, tenantId);
        // _tenantHolder.setTenantId(tenantId);
        Vertx.currentContext().putLocal(CoreUtils.KEY_TENANT_ID, tenantId);
        EPSettings settings = _coreStore.getTenantSettings(tenantId).getPaymentSettings();
        _gatewayFactory.getPaymentGateway(tenantId).start(tenantId, settings);
    }
```

and I have a tenant resolver that was initially written for HTTP requests, and which I modified to also check the Vertx context:
```
@ApplicationScoped
public class TenantResolver
{
    @Inject
    RoutingContext _routingContext;

    protected String tenantId()
    {
        Object value = null;
        try {
            Context context = Vertx.currentContext();
            value = context == null ? null : context.getLocal(CoreUtils.KEY_TENANT_ID);
        }
        catch (InjectionException e) {
            // Ignore and proceed with routing context.
        }

        if (value == null) {
            value = _routingContext.get(CoreUtils.KEY_TENANT_ID);
        }
        if (value == null) {
            throw new ServiceException(403, ErrorCode.TenantMissing);
        }
        return value.toString();
    }
}
```

I am not very clear about the `@ActivateRequestContext`, what context instances will that allow me to use? Trying to use _routingContext.put() in startGateway fails, I guess the context is only created for HTTP requests?

I tried using my own request scoped bean, but I don't know how to correctly access it from my tenant resolver:

```
@RequestScoped
public class TenantHolder
{
    private String _tenantId;
    public String getTenantId() { return _tenantId; }
    public void setTenantId(String tenantId) { _tenantId = tenantId; }
}
```

```
@ApplicationScoped
@Unremovable
public class EpTenantConfigResolver implements TenantConfigResolver
{
    @Inject
    TenantHolder _tenantHolder;

    @Override
    public Uni<OidcTenantConfig> resolve(RoutingContext context, OidcRequestContext<OidcTenantConfig> requestContext)
    {
        String tenantId = _getTargetTenant(context.request());
        if (tenantId != null) {
            // context.put(CoreUtils.KEY_TENANT_ID, tenantId);
            _tenantHolder.setTenantId(tenantId); -> fails since the resolver is ApplicationScoped
```

I could definitely use some pointers about the correct design here!
Thanks,
Franck

[a comment by an automated account is not shown]

**Comment by mkouba** [comment]
Hi @franck102, do you need to execute the `startGateway()` logic on the Vertx event loop? If so then the [`VertxContextSupport.subscribeAndAwait()`](https://quarkus.io/guides/vertx#executing-asynchronous-code-from-a-blocking-thread) util method might come in useful. It also starts the CDI request context if not already active.

> I am not very clear about the @ActivateRequestContext, what context instances will that allow me to use? Trying to use _routingContext.put() in startGateway fails, I guess the context is only created for HTTP requests?

Yes, the `io.vertx.ext.web.RoutingContext` is not available outside an HTTP request.

>  _tenantHolder.setTenantId(tenantId); -> fails since the resolver is ApplicationScoped

This should work just fine because the injected `_tenantHolder` is a [client proxy](https://quarkus.io/guides/cdi#client_proxies) that delegates invocations to the current instance of `TenantHolder` (if the request context is active).

**Reply by manovotn** [reply]
> I am not very clear about the @ActivateRequestContext, what context instances will that allow me to use? 

Just to add a clarification, the `@ActivateRequestContext` annotation is just a CDI annotation manually activating CDI request scope (for instance even outside of HTTP req.) and therefore allowing you to resolve and work with req. scoped beans. Without active context, you'd get `ContextNotActiveException`.
It isn't necessarily linked with vertx for which it seems you need actual HTTP request in this case - as per Martin's comment

**Reply by franck102** [reply]
> Hi @franck102, do you need to execute the `startGateway()` logic on the Vertx event loop? 

Hi @mkouba , the tasks are potentially long-running (calls to external payment providers) so if I read the Vertx docs correctly I should not run them on the event loop. And I am manually duplicating the current context so that I can use putLocal() since I am using the Vertx context currently.

I will try the TenantHolder approach again, I think I had the ActivateRequestContext annotation on the wrong method the first time.

**Reply by mkouba** [reply]
> Hi @mkouba , the tasks are potentially long-running (calls to external payment providers) so if I read the Vertx docs correctly I should not run them on the event loop.

Well, it depends whether you're using blocking or non-blocking APIs. If using blocking APIs then you should definitely not execute the logic on the event loop. 

Why exactly do you duplicate the vertx context? It's not a public API so 

Can't you just `@Inject Vertx` and then use `Vertx.executeBlocking()`? Do you even need to execute the logic on a Vertx context? 

Note that you can also `@Inject ExecutorService`, i.e. the default blocking executor, which is backed by Vertx worker threads (if Vertx extension is present).

**Reply by franck102** [reply]
> Why exactly do you duplicate the vertx context? It's not a public API so
> 
> Can't you just `@Inject Vertx` and then use `Vertx.executeBlocking()`? Do you even need to execute the logic on a Vertx context?
> 
> Note that you can also `@Inject ExecutorService`, i.e. the default blocking executor, which is backed by Vertx worker threads (if Vertx extension is present).

I was duplicating the context earlier to be able to use putLocal (fails on the root context) which is how I was passing the tenantId around.

Maybe I could use ExecutorService, but reading through quite a few dozens of posts / docs I couldn't figure out whether executeBlocking or the ExecutorService made more sense.
Then there is the ManagedExecutor too... all I know is I need context propagation (the tasks use the DB layer), and the tasks should run on worker threads.

**Reply by franck102** [reply]
> Note that you can also `@Inject ExecutorService`, i.e. the default blocking executor, which is backed by Vertx worker threads (if Vertx extension is present).

I just tried the default ManagedExecutor and injection of the RequestScoped bean is failing - it seems the ActivateRequestContext annotation has no effect.

```
 @Inject
    ManagedExecutor _executor;

    void onStart(@Observes StartupEvent ev)
    {
        for (String tenantId : _coreStore.listTenants()) {
            _executor.runAsync(() -> {
                try {
                    startGateway(tenantId);
                }
                catch (Throwable t) {
                    _log.error("Could not start payment gateway for tenant " + tenantId, t);
                }
            });
        }
    }

    @ActivateRequestContext
    public void startGateway(String tenantId)
    {
        _tenantHolder.setTenantId(tenantId);
```

The exception is
```
15:17:03 ERROR [co.pa.pa.se.PaymentService] (executor-thread-1) Could not start payment gateway for tenant totem: jakarta.enterprise.inject.IllegalProductException: Normal scoped producer method may not return null: io.quarkus.vertx.http.runtime.CurrentVertxRequest.getCurrent()
	at io.quarkus.vertx.http.runtime.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_Bean.doCreate(Unknown Source)
	at io.quarkus.vertx.http.runtime.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_Bean.create(Unknown Source)
	at io.quarkus.vertx.http.runtime.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.RequestContext.getIfActive(RequestContext.java:74)

**Reply by mkouba** [reply]
> The exception is...

Well, if you look at the error message it seems that it's not about inactive request context but the fact that something attempts to inject the `RoutingContext` which is not legal outside an HTTP request as explained earlier...

**Reply by franck102** [reply]
> > The exception is...
> 
> Well, if you look at the error message it seems that it's not about inactive request context but the fact that something attempts to inject the `RoutingContext` which is not legal outside an HTTP request as explained earlier...

You are right, TenantHolder.setTenantId() succeeds in startGateway, but when the TenantHolder is later injected in the task it's a different instance that hasn't been initialized.
getTenantId() seeing no explicit _tenantId defaults to using the RoutingContext which triggers the exception.

Using Vertx.executeBlocking I don't have that problem.

The task code is blocking and it is executed on a single thread.

**Comment by franck102** [comment]
I cannot use a RequestScoped bean in my OIDC TenantConfigResolver because the request context is not started yet when resolve() is called. So I am still putting the tenant on the RoutingContext there, then using this RequestScoped bean:
```
@RequestScoped
public class TenantHolder
{
    @Inject
    RoutingContext _routingContext;

    private String _tenantId;

    public String getTenantId()
    {
        return _tenantId != null ? _tenantId :
                _routingContext.get(CoreUtils.KEY_TENANT_ID);
    }

    public void setTenantId(String tenantId)  { _tenantId = tenantId;  }
}
```

With that I can simplify the Vertx code to use whatever context is available.
